Feat(cupertino): [date_picker] Date order parameter

ayshiff opened the pull request:
Closes #84550 

This PR adds a new `dateOrder` parameter to  `CupertinoDatePicker` to define the order of the columns in date mode which overrides the default order value defined by `localizations.datePickerDateOrder`.

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

ayshiff committed:
feat(cupertino): [date_picker] add date-order parameter

feat(cupertino): [date_picker] update docs

fix(cupertino): [date_picker] trailing space

fix(cupertino): [date_picker] remove unused line

fix(cupertino): [date_picker] format

pedromassango commented on packages/flutter/test/cupertino/date_picker_test.dart:
```diff
@@ -1221,6 +1221,33 @@ void main() {
+      lastOffset = tester.getTopLeft(find.text('14'));
+
+      expect(tester.getTopLeft(find.text('January')).dx > lastOffset.dx, true);
+      lastOffset = tester.getTopLeft(find.text('January'));
```
Is this last line required? I mean the assignment is not being used.

Maybe you meant to remove it or add another `expect` statement.

pedromassango reviewed (changes requested):
LGTM once the comments are addressed

pedromassango commented on packages/flutter/lib/src/cupertino/date_picker.dart:
```diff
@@ -1049,6 +1057,12 @@ class _CupertinoDatePickerDateTimeState extends State<CupertinoDatePicker> {
+
+  _CupertinoDatePickerDateState({
+    required this.dateOrder,
+  });
+  final DatePickerDateOrder? dateOrder;
```
Nit: space between

pedromassango reviewed (changes requested):
One more nit and should be go to go

pedromassango reviewed (approved):
LGTM

I would like to get a final approval from @LongCatIsLooong to make sure we're not missing anything.

LongCatIsLooong commented on packages/flutter/test/cupertino/date_picker_test.dart:
```diff
@@ -1221,6 +1221,32 @@ void main() {
+      expect(tester.getTopLeft(find.text('14')).dx > lastOffset.dx, true);
+      lastOffset = tester.getTopLeft(find.text('14'));
+
+      expect(tester.getTopLeft(find.text('January')).dx > lastOffset.dx, true);
```
```suggestion
      expect(
        tester.getTopLeft(find.text('14')).dx,
        lessThan(tester.getTopLeft(find.text('January')).dx,
      );
```

ayshiff committed:
fix(cupertino): [date_picker] update docs + tests

fix(cupertino): [date_picker] trailing space

fix(cupertino): [date_picker] update docs

fix(cupertino): [date_picker] update docs

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/date_picker.dart:
```diff
@@ -245,6 +245,9 @@ class CupertinoDatePicker extends StatefulWidget {
   /// [use24hFormat] decides whether 24 hour format is used. Defaults to false.
+  ///
+  /// [dateOrder] determines the order of the columns inside [CupertinoDatePicker] in date mode.
+  /// Defaults to [localizations.datePickerDateOrder].
```
`[localizations]` does not exist I think? Maybe just say defaults to the locale's default date format/order?

pedromassango commented:
@ayshiff could you please rebase for the CI to re-run?
